presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java

```diff
@@ -148,9 +149,11 @@ public boolean next()
         try {
             if (!results.hasNext()) {
                 row.set(null);
+                currentRow = 0;
                 return false;
             }
             row.set(results.next());
+            currentRow++;
```

Comment:
What should be the expected behavior after reading 2^31 rows?

Reply:
As far as I checked quickly (not tested) other JDBC drivers, they will just overflow. 

PostgreSQL: https://github.com/pgjdbc/pgjdbc/blob/c80add9975d88a134319374cf5dd4adec864f432/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L741

MySQL: https://github.com/mysql/mysql-connector-j/blob/d64b664fa93e81296a377de031b8123a67e6def2/src/main/user-impl/java/com/mysql/cj/jdbc/result/ResultSetImpl.java#L1560

SQLServer: https://github.com/microsoft/mssql-jdbc/blob/606e722fbee129a4fd0de89423f5a934207522b0/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResultSet.java#L1494

Reply:
Thanks! So should I detect the overflow and define the expected behavior? If so, probably I should throw exception for that case. It seems like I can only throw SQLException but per the spec, this exception is threw only when if a database access error occurs or this method is called on a closed result set.

Reply:
Hi, @findepi @ebyhr, may I ask if there is any suggestion on this? Thanks!

Reply:
Given that this is unspecified behavior in the JDBC spec, and that none of the common JDBC implementations seem to handle this, my guess is that this is not too big a deal either way. If we want to be really pedantic about this, it would probably be just slightly safer to just throw a SQLException in this case because a negative row number returned is not a valid row number. Given that JDBC does not permit more than Integer.MAX_VALUE rows to be returned in one shot, you could argue this is a DB access error.

Reply:
Thanks for your suggestion! That makes sense to me.

Reply:
Note that we cannot throw upon increment (this would break anyone reading more than 2B rows, which is possible today, if you're patient enough).
For this, we can keep the counter as a long (AtomicInteger->AtomicLong) and verify the value is within `[0, Integer.MAX_VAULUE]` within `getRow()` method.
(long counter won't overflow in practice)

Reply:
@findepi  OK, so do you mean we allow it overflow during increment (e.g. when calling `next`). However, when `getRow` is called, we need to do overflow verification and throw exception if it is overflow?

Reply:
@jasonyanwenl, he means to use a long for the internal row counter so that it doesn't actually overflow, but just to throw the exception if someone tries to fetch the the current row that can't be squeezed into the JDBC API int return value.

Reply:
OK. Got it!

Reply:
@jasonyanwenl at first there seems to be no much difference between int and long.
in practise however, it will take you thousands years to overflow long even if you increase the counter at 1M/s pace.